File: Reorder List/Reorder_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
#include <iostream>
#include <vector>

typedef struct ListNode
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
} ListNode;

class Solution
{
public:
    void reorderList(ListNode *head)
    {
        if (!head || !head->next)
            return;

        ListNode *slow = head, *fast = head;
        while (fast && fast->next) //start to split it into two halves, the right should be less than the left;
        {
            slow = slow->next;
            fast = fast->next->next;
        }

        ListNode *prev = NULL, *node = slow, *next = NULL;
        while (node)
        {
            next = node->next;
            node->next = prev;
            prev = node;
            node = next;
        }

        ListNode *first = head, *second = prev;
        while (second->next)
        {
            ListNode *pfirst = first->next;
            ListNode *psecond = second->next;
            first->next = second;
            first = pfirst;
            second->next = first;
            second = psecond;
        }
        return;
    }
};

int main()
{
    Solution a;

    std::vector<int> arr{2, 3, 4, 5, 6, 7, 8};
    ListNode *head = new ListNode(1);
    ListNode *p = head;
    // p = new ListNode(1);
    for (auto i : arr)
    {
        // std::cout << "i = " << i << "\n";
        ListNode *next = new ListNode(i);
        p->next = next;
        // std::cout << "value = " <<p->val << "\n";
        p = next;
    }

    a.reorderList(head);

    ListNode *s1 = head;
    ListNode *s2 = s1;
    while (s2!= nullptr)
    {
        std::cout << s1->val << " ";
        s1 = s2->next;
        delete s2;
        s2 = s1;
        // if (s2 == nullptr) std::cout << "\n";
    }
    s1 = s2 = nullptr;

    return 0;
}

File: Reorder_List/Reorder_List.py
# https://leetcode.com/problems/reorder-list/
# Definition for singly-linked list.
'''
Given a singly linked list L: L0→L1→…→Ln-1→Ln,
reorder it to: L0→Ln→L1→Ln-1→L2→Ln-2→…

You may not modify the values in the list's nodes, only nodes itself may be changed.
Example 1:

Given 1->2->3->4, reorder it to 1->4->2->3.
Example 2:

Given 1->2->3->4->5, reorder it to 1->5->2->4->3.
'''


class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
    def reorderList(self, head):
        """
        :type head: ListNode
        :rtype: None Do not return anything, modify head in-place instead.
        """
        if not head:
            return None
        # set slow to mid for odd length, first of second half for even
        fast, slow = head, head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
        
        prev, node = None, slow
        while node:
            prev, node.next, node = node, prev, node.next

        first, second = head, prev

        while second.next:
            first.next, first = second, first.next
            second.next, second = first, second.next


if __name__ == "__main__":
    head = ListNode(1)
    next_node = head
    l = [2, 3, 4, 5, 6, 7, 8]

    for i in l:
        node = ListNode(i)
        next_node.next = node
        next_node = next_node.next

    n = head
    while n:
        print(n.val, end=" ")
        n = n.next

    print('')
    print("reorder list")
    s = Solution()
    s.reorderList(head)

    while head:
        print(head.val, end=" ")
        head = head.next
